core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+        return value;
+      }
+
+      auto encodeUInt8(uint8_t value) {
+        return value;
+      }
+
+      auto encodeInt16(uint16_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeUint16(uint16_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeInt32(uint32_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeUint32(uint32_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeInt64(int64_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto encodeUint64(uint64_t value) {
+        return encodePlainInteger(value);
+      }
+
+      auto decodeInt8(int8_t value) {
+        return value;
+      }
+
+      auto decodeUInt8(uint8_t value) {
+        return value;
+      }
+
+      auto decodeInt16(uint16_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeUint16(uint16_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeInt32(uint32_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeUint32(uint32_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeInt64(int64_t value) {
+        return decodePlainInteger(value);
+      }
+
+      auto decodeUint64(uint64_t value) {
+        return decodePlainInteger(value);
+      }
+
+      uint8_t encodeBool(bool value) {
+        return value ? static_cast<uint8_t>(0x01) : static_cast<uint8_t>(0x00);
+      }
+
+      DecodeBoolResult decodeBool(uint8_t byte) {
+        if (byte == 0) {
+          return expected::Value{false};
+        } else if (byte == 1) {
+          return expected::Value{true};
+        }
+        return expected::Error{DecodeBoolError::kUnexpectedValue};
+      }
+    }  // namespace fixedwidth
+
+    namespace compact {
+      struct EncodingCategoryLimits {
+        enum type : size_t {
+          kLessThan64 = 1ul << 6,
+          kLessThan16384 = 1ul << 14,
+          kLessThanTwoDegreeThirty = 1ul << 30
+        };
+      };
+
+      template <class T>
+      struct EncodedLengthCalculator {
+        EncodedLengthCalculator() {}
+
+        size_t operator()(const T &) {
+          assert(false && "not implemented for this type");
+          return 0;
+        }
+      };
+
+      EncodeIntegerResult encodeUint8(uint8_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThan64) {
+          const uint8_t v = (value << 2);
+          return expected::Value{ByteArray{v}};
+        }
+
+        return encodeUint16(static_cast<uint16_t>(value));
+      }
+
+      EncodeIntegerResult encodeInt8(int8_t value) noexcept {
+        return encodeUint8(util::makeSignReinterpreter(value).asUnsigned());
+      }
+
+      EncodeIntegerResult encodeUint16(uint16_t value) noexcept {
+        if (value < EncodingCategoryLimits::kLessThan16384) {
+          uint16_t v = value;
+          v <<= 2;  // v *= 4
```

Comment:
compiler is smart enough to optimize `v = v * 4 + 1`, but the source code is still for humans

Reply:
In my opinion v = v*4 + 1 is less obvious than explanation how the value is formed.
Don't you think so?

Reply:
Hm, what useful info `v <<= 2` statement provides?

Reply:
It says that bits representing number are moved left.

Reply:
Don't you mind if I keep it as is?